### PR TITLE
Fix “Git is unavailable” error.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

### CODE REVIEW

**Comment by reitermarkus** [comment]
I think I found the bug.

`OS::Mac` a.k.a. `MacOS` is undefined when running tests with `--generic`, so installing `git` fails because the `pkg-config` dependency uses `MacOS`.

**Comment by MikeMcQuaid** [comment]
> `OS::Mac` a.k.a. `MacOS` is undefined when running tests with `--generic`, so installing `git` fails because the `pkg-config` dependency uses `MacOS`.

Any ideas why this only failed after refactoring?

File: Library/Homebrew/utils/git.rb [hunk 1]
```diff
@@ -93,8 +93,14 @@ def ensure_installed!
       if CoreTap.instance.installed?
         begin
           oh1 "Installing #{Formatter.identifier("git")}"
-          safe_system HOMEBREW_BREW_FILE, "install", "git"
-          clear_available_cache
+
+          # We need to unset `HOMEBREW_TEST_GENERIC_OS`, otherwise `git` will be
+          # installed from source in tests that need it. This is slow and will
+          # also likely fail due to `OS::Linux` and `OS::Mac` being undefined.
+          with_env "HOMEBREW_TEST_GENERIC_OS" => nil do
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/git.rb:100]: This is a recent regression and warrants a better fix. Why was Git not being seen as available when it was?

**Comment by reitermarkus** [comment]
> Any ideas why this only failed after refactoring?

Previously `||=` was used for caching instead of `defined?`, so `available = false` wasn't actually cached.

**Comment by MikeMcQuaid** [comment]
Why is it showing up as unavailable with tests running on GitHub Actions where it is available? `HOMEBREW_TEST_GENERIC_OS` leaking into the code is a hack around this functionality not working correctly.

**Comment by reitermarkus** [comment]
> `HOMEBREW_TEST_GENERIC_OS` leaking into the code is a hack

I think this is still technically correct, even if the test failure is a regression, since calling `Git.ensure_installed!` with `--generic` and `git` not installed would result in this error.

**Comment by MikeMcQuaid** [comment]
> I think this is still technically correct, even if the test failure is a regression, since calling `Git.ensure_installed!` with `--generic` and `git` not installed would result in this error.

That logic is reliant on the formula, though, no, and technically there are no "generic" formulae. I'd remove it for now in https://github.com/Homebrew/brew/pull/8496

**Pull request merged**